elastic/beats
# Add support for port mapping in docker hints

## What does this PR do?
Add support for port mapping in docker autodiscover. The idea is similar to what we do for Kubernetes which was added at https://github.com/elastic/beats/pull/19398. 

The idea is to make it possible for collecting metrics using the HostPort of the container as described at https://github.com/elastic/beats/issues/19825, by referring to the proper internal port of the container (see manual testing notes for an example).

## Why is it important?
To cover cases like https://github.com/elastic/beats/issues/19825.


## How to test this PR locally

1. Use the following configuration:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      hints.enabled: true

output.console:
  pretty: true
```
2. Run metricbet with `./metricbeat -e -d "module" -c metricbeat.yml`
3. Start a target container with `docker run --name nats --rm -p 4222:4222 -p 8081:8222 --label co.elastic.metrics/module="nats" --label co.elastic.metrics/metricsets="stats" --label co.elastic.metrics/hosts='localhost:${data.ports.8222}' nats:2.1.4 --http_port 8222`
4. Verify in the console output that Metricbeat is collecting metrics properly from `localhost:8222` endpoint.

## Related issues

- Closes https://github.com/elastic/beats/issues/19825
- Relates https://github.com/elastic/beats/pull/19398
